From 4b56d40a9989a7d74e057023997a54309b58e39c Mon Sep 17 00:00:00 2001
From: Avery Pennarun <apenwarr@gmail.com>
Date: Thu, 13 Feb 2014 05:12:04 -0500
Subject: [PATCH 1/2] mwifiex: don't free the adapter from the firmware loader.

It's already being freed elsewhere, so this results in a double free.
---
 drivers/net/wireless/mwifiex/main.c | 17 ++++++++++-------
 1 file changed, 10 insertions(+), 7 deletions(-)

diff --git a/drivers/net/wireless/mwifiex/main.c b/drivers/net/wireless/mwifiex/main.c
index 5b6df5d..51638ca 100644
--- a/drivers/net/wireless/mwifiex/main.c
+++ b/drivers/net/wireless/mwifiex/main.c
@@ -397,9 +397,11 @@ static void mwifiex_free_adapter(struct mwifiex_adapter *adapter)
  */
 static void mwifiex_terminate_workqueue(struct mwifiex_adapter *adapter)
 {
-	flush_workqueue(adapter->workqueue);
-	destroy_workqueue(adapter->workqueue);
-	adapter->workqueue = NULL;
+	if (adapter->workqueue) {
+		flush_workqueue(adapter->workqueue);
+		destroy_workqueue(adapter->workqueue);
+		adapter->workqueue = NULL;
+	}
 }
 
 /*
@@ -521,8 +523,6 @@ done:
 		adapter->firmware = NULL;
 	}
 	complete(&adapter->fw_load);
-	if (init_failed)
-		mwifiex_free_adapter(adapter);
 	up(sem);
 	return;
 }
@@ -998,8 +998,11 @@ int mwifiex_remove_card(struct mwifiex_adapter *adapter, struct semaphore *sem)
 		rtnl_unlock();
 	}
 
-	wiphy_unregister(adapter->wiphy);
-	wiphy_free(adapter->wiphy);
+	if (adapter->wiphy) {
+		wiphy_unregister(adapter->wiphy);
+		wiphy_free(adapter->wiphy);
+		adapter->wiphy = NULL;
+	}
 
 	mwifiex_terminate_workqueue(adapter);
 
-- 
1.9.0.rc1.175.g0b1dcb5

